modules/apps/transfer/types/v3/packet_test.go

```diff
@@ -0,0 +1,571 @@
+				sender,
+				receiver,
+				fmt.Sprintf(`{"src_callback": {"address": "%s", "gas_limit": "200000"}}`, receiver)),
+			map[string]interface{}{
+				"address":   receiver,
+				"gas_limit": "200000",
+			},
+		},
+		{
+			"success: src_callback has string value",
+			NewFungibleTokenPacketData(
+				[]*Token{
+					{
+						Denom:  "atom/pool",
+						Amount: "1000",
+						Trace:  []string{"transfer/channel-0", "transfer/channel-1"},
+					},
+				},
+				sender,
+				receiver,
+				`{"src_callback": "string"}`),
+			"string",
+		},
+		{
+			"failure: src_callback key not found memo",
+			NewFungibleTokenPacketData(
+				[]*Token{
+					{
+						Denom:  "atom/pool",
+						Amount: "1000",
+						Trace:  []string{"transfer/channel-0", "transfer/channel-1"},
+					},
+				},
+				sender,
+				receiver,
+				fmt.Sprintf(`{"dest_callback": {"address": "%s", "min_gas": "200000"}}`, receiver)),
+			nil,
+		},
+		{
+			"failure: empty memo",
+			NewFungibleTokenPacketData(
+				[]*Token{
+					{
+						Denom:  "atom/pool",
+						Amount: "1000",
+						Trace:  []string{"transfer/channel-0", "transfer/channel-1"},
+					},
+				},
+				sender,
+				receiver,
+				""),
+			nil,
+		},
+		{
+			"failure: non-json memo",
+			NewFungibleTokenPacketData(
+				[]*Token{
+					{
+						Denom:  "atom/pool",
+						Amount: "1000",
+						Trace:  []string{"transfer/channel-0", "transfer/channel-1"},
+					},
+				},
+				sender,
+				receiver,
+				"invalid"),
+			nil,
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+
+		customData := tc.packetData.GetCustomPacketData("src_callback")
+		require.Equal(t, tc.expCustomData, customData)
+	}
+}
+
+func TestFungibleTokenPacketDataOmitEmpty(t *testing.T) {
+	testCases := []struct {
+		name       string
+		packetData FungibleTokenPacketData
+		expMemo    bool
+	}{
+		{
+			"empty memo field, resulting marshalled bytes should not contain the memo field",
+			NewFungibleTokenPacketData(
+				[]*Token{
+					{
+						Denom:  "atom/pool",
+						Amount: "1000",
+						Trace:  []string{"transfer/channel-0", "transfer/channel-1"},
+					},
+				},
+				sender,
+				receiver,
+				"",
+			),
+			false,
+		},
+		{
+			"non-empty memo field, resulting marshalled bytes should contain the memo field",
+			NewFungibleTokenPacketData(
+				[]*Token{
+					{
+						Denom:  "atom/pool",
+						Amount: "1000",
+						Trace:  []string{"transfer/channel-0", "transfer/channel-1"},
+					},
+				},
+				sender,
+				receiver,
+				"abc",
+			),
+			true,
+		},
+	}
+
+	for _, tc := range testCases {
+		bz, err := json.Marshal(tc.packetData)
+		if tc.expMemo {
+			require.NoError(t, err, tc.name)
+			// check that the memo field is present in the marshalled bytes
+			require.Contains(t, string(bz), "memo")
+		} else {
+			require.NoError(t, err, tc.name)
+			// check that the memo field is not present in the marshalled bytes
+			require.NotContains(t, string(bz), "memo")
+		}
+	}
+}
+
+func TestGetFullDenomPath(t *testing.T) {
```

Comment:
This function and `TestValidate` below could go to token_test.go if the functions of the `Token` type move to their own file.